### PR TITLE
std.sync.atomic: extended atomic helper functions

### PR DESCRIPTION
**Pull request by kprotty**
* generic Atomic(T) wrapper replacing atomic.Int and atomic.Bool
* Bitwise atomic intrinsics when Atomic(T) is an Integer
* No redundant specification of type as seen in builtins atomic functions
* fetchXX() ops instead of generic atomicRmw for C11 consistency 
* contains compilerFence() and spinLoopHint()

### CODE REVIEW

File: lib/std/sync/atomic.zig [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+
+const std = @import("../std.zig");
+const target = std.Target.current;
+const testing = std.testing;
+
+const AtomicOrder = std.builtin.AtomicOrder;
+const AtomicRmwOp = std.builtin.AtomicRmwOp;
+const AtomicBitRmwOp = enum { Set, Reset, Toggle };
+
+pub fn spinLoopHint() void {
```
Comment (LemonBoy) [inline lib/std/sync/atomic.zig:15]: There's already a copy of this function in the stdlib with several enhancements.

**Comment by LemonBoy** [comment]
We already have atomic wrappers for ints and bools. That approach is much safer than taking raw pointers as it prevents you from mixing atomic and non-atomic ops.

File: lib/std/sync/atomic.zig (hunk 1 above)
Comment (kprotty) [reply lib/std/sync/atomic.zig:15]: Just saw the enhanced version. Will move `std.Thread.spinLoopHint()` here and have it alias this instead

**Comment by kprotty** [comment]
@LemonBoy atomic int and bool dont contain cmpxchg abstractions + there aren't any such wrappers for atomic pointers/enums/floats. Not sure about improved safety as ordering misuse/ignorance can often be equally unsafe. Mixing non-atomic ops and atomic ops can sometimes be valid. Would a generic `Atomic(T)` type with `load/storeUnchecked()` and limiting fetchXX()/bitXX() for when T is an Int be a good compromise?

**Comment by LemonBoy** [comment]
> atomic int and bool dont contain cmpxchg abstractions 

Should be added if needed.

> there aren't any such wrappers for atomic pointers/enums/floats

Wrappers for pointers and enums are also a good idea. Floating point types can be added later (if ever) when somebody finds an use for those.

Not sure about improved safety as ordering misuse/ignorance can often be equally unsafe.

The wrappers try to protect the user from shooting themselves in the foot by checking the ordering parameter, avoid defining meaningless operations, and preventing the mix-up of atomic & non atomic ops. The idea of wrapper types is not novel (see Rust or C11) and strikes a good balance between ergonomics and safety.

> Would a generic Atomic(T)

I think separate wrapper types are simpler to define, you need quite a bit of meta programming and `usingnamespace` to achieve the same effect and define some helpers only when they make sense.

**Comment by kprotty** [comment]
Separate wrapper types feels like repetition: They would all expose load/store/xchg/cmpxchg and ordering checks. Floats would expose that plus arithmetic rmw. Ints would expose even that plus bitwise rmw. Theres few metaprogramming conditions but it covers a lot of cases. Thats my pitch for Atomic(T)

**Comment by LemonBoy** [comment]
> Thats my pitch for Atomic(T)

I'm sold!
Given we're in for some sweeping changes, what do you think of using the same nomenclature as the [C++](https://en.cppreference.com/w/cpp/atomic) atomic types (`load, store, exchange, fetch{Add,Sub,...}, compareExchange{Strong,Weak},...`) ?

**Comment by kprotty** [comment]
I prefer everything except the `exchange` bits - replacing them for `swap`/`[try]compareAndSwap` is on average shorter and is reminiscent of [Go's atomic naming scheme](https://golang.org/pkg/sync/atomic/).

Actually, another issue with wrapper types is being able to store them inside `extern struct/union`. Could store the generic bytes and cast between them like `atomic_bytes: [@sizeOf(Atomic(T))] align(@alignOf(Atomic(T))` but that is pretty hacky.

**Comment by LemonBoy** [comment]
> Actually, another issue with wrapper types is being able to store them inside extern struct/union

There should be no problem if the wrapper is `extern`.

**Comment by kprotty** [comment]
freebsd CI: `atomic.zig:25:13: error: expected 0 parameters, found 1`

atomic.zig:25: `@fence(ordering)`

Whats up with the fence builtin?

**Comment by LemonBoy** [comment]
Change [this](https://github.com/ziglang/zig/blob/6435750c99e705eb40bbdf75e51a3493d683e951/src/BuiltinFn.zig#L403) line.

File: lib/std/atomic.zig [hunk 2]
```diff
@@ -3,14 +3,88 @@
...
+        else => {
+            @compileLog(ordering, " only applies to a given memory location");
+        },
+    }
+}
+
+pub fn compilerFence(comptime ordering: Ordering) callconv(.Inline) void {
+    switch (ordering) {
+        .SeqCst => {
+            // TODO: Use LLVM intrinsic `fence syncscope("singlethread")` instead of inline asm.
```
Comment (LemonBoy) [inline lib/std/atomic.zig:36]: There are some [unresolved problems](https://github.com/rust-lang/rust/issues/62256) with the use of singlethread fences, the biggest one is that they may produce extra instructions or a call (!) to `__sync_synchronize`.
Let's use the good old empty-asm-block trick for the time being (and group the ordering parameters, having to expand 3 layers of recursion is pretty bad for stage1, beside being hard to understand).

Comment (kprotty) [reply lib/std/atomic.zig]: Weird find.. 

fixed it

File: lib/std/atomic/Atomic.zig [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+        var x = Atomic(usize).init(5);
+        try testing.expectEqual(x.swap(10, ordering), 5);
+        try testing.expectEqual(x.load(.SeqCst), 10);
+
+        var y = Atomic(enum(usize) { a, b, c }).init(.c);
+        try testing.expectEqual(y.swap(.a, ordering), .c);
+        try testing.expectEqual(y.load(.SeqCst), .a);
+
+        var z = Atomic(f32).init(5.0);
+        try testing.expectEqual(z.swap(10.0, ordering), 5.0);
```
Comment (LemonBoy) [inline lib/std/atomic/Atomic.zig:279]: Skip the `f32` tests for AAarch64 or the build will fail with `LLVM ERROR: Cannot select: intrinsic %llvm.aarch64.stxr`.

File: lib/std/atomic.zig [hunk 4]
```diff
@@ -3,14 +3,82 @@
...
+
+        // `isb` appears more reliable for releasing execution resources than `yield` on common aarch64 CPUs.
+        // https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8258604
+        // https://bugs.mysql.com/bug.php?id=100664
+        .aarch64, .aarch64_be, .aarch64_32 => "isb",
+
+        // `yield` was introduced in v6k but is also available on v6m.
+        // https://www.keil.com/support/man/docs/armasm/armasm_dom1361289926796.htm
+        .arm, .armeb, .thumb, .thumbeb => blk: {
+            const can_yield = std.Target.arm.featureSetHasAny(target.cpu.features, .{ .has_v6k, .has_v6m });
```
Comment (LemonBoy) [inline lib/std/atomic.zig:66]:
```suggestion
            const can_yield = comptime std.Target.arm.featureSetHasAny(target.cpu.features, .{ .has_v6k, .has_v6m });
```

File: lib/std/atomic/Atomic.zig (hunk 3 above)
Comment (LemonBoy) [reply lib/std/atomic/Atomic.zig]: Now that #8922 is merged you can remove the AArch64 check, the tests should pass on both that and ARM now.

**Comment by kprotty** [comment]
@LemonBoy Any further comments before it gets merged?

File: lib/std/atomic/Atomic.zig [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+            compare: T,
+            exchange: T,
+            comptime success: Ordering,
+            comptime failure: Ordering,
+        ) callconv(.Inline) ?T {
+            if (success == .Unordered or failure == .Unordered) {
+                @compileError(@tagName(Ordering.Unordered) ++ " is only allowed on atomic loads and stores");
+            }
+
+            comptime var success_is_stronger = switch (failure) {
```
Comment (LemonBoy) [inline lib/std/atomic/Atomic.zig:87]:
```suggestion
            comptime const success_is_stronger = switch (failure) {
```

**Review by LemonBoy — APPROVED** [review]
LGTM.
The size of type `T` is actually bounded by the combination of architecture and feature flags being used, an extra check can be added at a later stage though.

**Pull request merged**